server/user.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * @fileoverview user function handler
+ */
+var fs = require('fs');
+var assert = require('assert');
+
+var log = require('./log.js');
+var utils = require('./utils.js');
+
+/** @type {user} */
+module.exports = user;
+
+/**
+ * @constructor
+ */
+function user() {}
+
+/** @enum {string} */
+user.QueryType = {
+  SIGNUP: 'sign-up',
+  SIGNIN: 'sign-in'
+};
+
+/**
+ * Expire time of cookie.
+ * @const {number}
+ */
+user.cookieExpireTime = 24 * 60 * 60 * 1000;
+
+/**
+ * @typedef {{
+ *   email: (string|undefined),
+ *   username: string,
+ *   password: string,
+ *   confirmed: (boolean|undefined)
+ * }}
+ */
+user.Info;
+
+/**
+ * @typedef {{
+ *   error: string
+ * }}
+ */
+user.Error;
+
+/**
+ * Checks the user information for signing in.
+ * @param {!mongodb.Db} db Session database.
+ * @param {string} userPath Directory of user information file.
+ * @param {!user.Info} userInfo User Information
+ * @param {function(!Object)} callback Callback function.
+ */
+user.signUp = function(db, userPath, userInfo, callback) {
+  var checkDuplicate = {
+    duplicated: false,
+    elements: []
+  };
+  var cursor = db.collection('userInfo').find({
+    $or: [{username: userInfo.username}, {email: userInfo.email}]});
+  var data = [];
+  cursor.each(function(err, doc) {
+    assert.equal(err, null, 'error occurs');
+    if (doc != null) {
+      data.push(doc);
+    } else {
+      var result = authenticateCallback();
+      callback(result);
+    }
+  });
+  var authenticateCallback = function() {
```

Comment:
Writing nested functions bypasses the compiler's check of fucntion definitions. Besides too many nested functions harm the code readability as it increases the code structure complexity. Typically only light-weight helper functions are nested. Your nested function also accesses outer scoped variable "data", which is a code style that is not recommended and can be avoided in this case.

It seems that this callback is pretty large in size, and could be written as
```js
genotet.user.signupUser = function(userInfo) {
}
```

IMO "will not be used in any other places" is not a strong enough motivation to create nested functions. Besides, functions just represent a procedure and are not necessarily have to be "generic". For example, many functions in renderers are not used in any other places either. The code architecture right now splits different namespaces into different files. Preferrably each file shall then contain flattened, simple-form functions.

Additionally the program flow here is a bit unclear. The db call just looks for duplicates. Then I think a better structure is to validate duplicates in your db callback, then call signupUser once the signup info has been validated, so that signupUser can assume userInfo is valid and just perform db updates.